proposals/webidl-bindings/Explainer.md

```diff
@@ -0,0 +1,623 @@
+* If a JavaScript Built-in function that today has no Web IDL signature is
+  imbued with a Web IDL signature by a future iteration of the spec, this new
+  Web IDL signature may be subtly incompatible with extant uses of the
+  Built-in; the fallback behavior ensures adding a Web IDL signature to a
+  Built-in is always backwards-compatible.
+* If two browsers have slightly Web IDL signatures (which can occur for
+  historical compatibility reasons); the fallback behavior helps avoid
+  unnecssary hard failures.
+
+(To help well-intentioned developers avoid unintended performance degredation,
+WebAssembly engines should emit warning diagnostics on Web IDL signature
+mismatch.)
+
+
+## Example: TextEncoder.encodeInto
+
+An abridged Web IDL signature for [`encodeInto`] is:
+
+```
+dictionary TextEncoderEncodeIntoResult {
+  unsigned long long read;
+  unsigned long long written;
+};
+
+[Constructor]
+interface TextEncoder {
+  TextEncoderEncodeIntoResult encodeInto(USVString source, Uint8Array destination);
+};
+```
+
+The signature of `encodeInto` could be defined with these (strawman text-format)
+statements:
+
+```
+(webidl-type $TEEIR (dictionary (field "read" unsigned long long) (field "written" unsigned long long)))
+(webidl-type $EncodeIntoFuncWebIDL (func (method any) (param USVString Uint8Array) (result $TEEIR)))
+```
+
+The WebAssembly signature and import which would be used to call `encodeInto`
+would be:
+
+```
+(type $EncodeIntoFuncWasm (param anyref anyref i32 i32) (result i64 i64))
+(func $encodeInto (import "TextEncoder" "encodeInto") (type $EncodeIntoFuncWasm)
+```
+
+Finally, the binding is accomplished with these two (strawman text-format)
+statements:
+
+```
+(webidl-func-binding $encodeIntoBinding import $EncodeIntoFuncWasm $EncodeIntoFuncWebIDL
+    (param (as type=any idx=0) (as type=any idx=1) (view type=uint8 off-idx=2 len-idx=3))
+    (result (as type=i64 (field "read" (get 0))) (as type=i64 (field "written" (get 0)))))
+(webidl-bind $encodeInto $encodeIntoBinding)
+```
+
+This function binding can be visualized:
+
+![encodeInto Binding Diagram](example-encodeInto-binding.svg)
+
+To get the `anyref` `this`-value, a WebAssembly module must *additionally*
+import the `TextEncoder` constructor, with the strawman
+`(constructor default-new-target)` syntax indicating the default `newTarget`
+(the callee constructor).
+
+```
+(webidl-type $CtorFuncWebIDL (func (constructor default-new-target) (result any)))
+
+(type $CtorFuncWasm (func (result anyref)))
+(func $TextEncoderCtor (import "TextEncoder" "constructor") (type $CtorFuncWasm))
+
+(webidl-func-binding $CtorBinding import $CtorFuncWasm $CtorFuncWebIDL
+  (result (as type=anyref (get 0))))
+(webidl-bind $TextEncoderCtor $CtorBinding)
+```
+
+With these bindings, a WebAssembly module can call `$TextEncoderCtor`, store the result in an `anyref` global and then pass that global value as the first argument to all subsequent calls to `$encodeInto`.
```

Comment:
To tie everything together as well, could this include an exapmle of that the `importObject` to `WebAssembly.instantiate` would be to cover `TextEncoderCtor` and  `encodeInto`?

Reply:
Great idea, done!